### PR TITLE
[melange]: Remove `(target DIR)` from `melange.emit`

### PR DESCRIPTION
**Pull request by anmonteiro**
Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

### CODE REVIEW

**Comment by rgrinberg** [comment]
Okay, I pushed a little hack that gives us the machinery in the engine. Now we just need to test it by modifying gen_rules to use this new construct whenever we're under  the`melange.emit` stanza

**Comment by anmonteiro** [comment]
This should be ready for review. I need help understanding why the `merlin.t` test doesn't include the `Main` module anymore. I broke something.

File: test/blackbox-tests/test-cases/melange/intfonly.t/run.t [hunk 1]
```diff
@@ -1,6 +1,20 @@
 Test melange libs flow when using `modules_without_implementation` stanza
 
 Build js files
-  $ dune build output/b.js
-  $ node _build/default/output/b.js
+  $ dune build @dist
+
+writing a `melange.emit` stanza in the root `dune` file overlays JS files with
+the source files (sort added because GNU find and BSD find have different
```
Comment (jchavarri) [inline test/blackbox-tests/test-cases/melange/intfonly.t/run.t:7]: Maybe use `tree` instead?

**Review by jchavarri — COMMENTED** [review]
Looks great. I would feel safer if there were some tests added for the following cases, which become more relevant after the changes in the PR:
- `melange.emit` using `libraries` placed in ancestor folders
- two or more `melange.emit` stanzas in the same folder

File: test/blackbox-tests/test-cases/melange/intfonly.t/run.t (hunk 1 above)
Comment (anmonteiro) [reply test/blackbox-tests/test-cases/melange/intfonly.t/run.t:7]: `tree` isn't used anywhere in the tests, whereas there are multiple instances of `find`.

**Comment by anmonteiro** [comment]
> melange.emit using libraries placed in ancestor folders

will add a test for this

> two or more melange.emit stanzas in the same folder

this is currently not supported (see [this piece of code](https://github.com/ocaml/dune/blob/5a098272172c1f16aad309d6aabc016f7be2a9f3/src/dune_rules/ml_sources.ml#LL76C20-L76C20)). I'm actually thinking about relaxing this assumption if the `javascript_extension` is different. But that's a separate PR. mind if I address it then?

**Comment by anmonteiro** [comment]
@jchavarri https://github.com/ocaml/dune/pull/6541/commits/889075db414e507e4fed32eacff77bb032636dd9 adds a nested emit test case

**Comment by jchavarri** [comment]
@anmonteiro There was only a small failure in `merlin` test that I fixed in https://github.com/anmonteiro/dune/pull/1 (can merge that PR into this one).

Besides that, is there anything else missing for the PR to be reviewed by @rgrinberg ?

**Comment by anmonteiro** [comment]
Feel free to merge. I can rebase this until end of week.

**Comment by jchavarri** [comment]
I don't think I have permissions to merge my PR or push to this branch, so deferring to @rgrinberg to see if something can be done.

**Comment by anmonteiro** [comment]
Merged. Just need to rebase later in the week.

**Comment by anmonteiro** [comment]
we'll try to explore this again when the masking RFC is implemented.

**Pull request closed without merging**